core/services.go

```diff
@@ -703,6 +703,34 @@ func (app *Application) updatePost(clientID string, current *model.User, post *m
 	return app.storage.UpdatePost(clientID, current.ID, post)
 }
 
+func (app *Application) reactToPost(clientID string, current *model.User, groupID string, postID string, reaction string) error {
+	post, err := app.storage.FindPost(clientID, &current.ID, groupID, postID, true, false)
+	if err != nil {
+		return fmt.Errorf("error finding post: %v", err)
+	}
+	if post == nil {
+		return fmt.Errorf("missing post for id %s", postID)
+	}
+
+	for _, userID := range post.Reactions[reaction] {
+		if current.ID == userID {
+			err = app.storage.ReactToPost(current.ID, postID, reaction, false)
+			if err != nil {
+				return fmt.Errorf("error removing reaction: %v", err)
+			}
+
+			return nil
+		}
+	}
+
+	err = app.storage.ReactToPost(current.ID, postID, reaction, true)
```

Comment:
These need to use transactions to avoid potential duplication as a result of race conditions

Reply:
Addressed by #249